src/NHibernate.Test/NHSpecificTest/GH2067/BaseClassAndDerivedInterfaceProxiesFixture.cs

```diff
@@ -0,0 +1,113 @@
+using System;
+using NHibernate.Cfg.MappingSchema;
+using NHibernate.Mapping.ByCode;
+using NUnit.Framework;
+
+namespace NHibernate.Test.NHSpecificTest.GH2067
+{
+	namespace BaseClassAndDerivedInterfaceProxies
+	{
+		//Base [Base] -> Derived1 [IDerived1] -> Derived2 [IDerived2]
+		public class NotMappedBase
+		{
+		}
+
+		public class Base : NotMappedBase
+		{
+			public virtual Guid Id { get; set; }
+			public virtual int BaseName { get; set; }
+		}
+
+		public class Derived1Interface : Base, IDerived1
+		{
+			public virtual string Derived1Name { get; set; }
+		}
+
+		public class Derived2Interface : Derived1Interface, IDerived2
+		{
+			public virtual string Derived2Name { get; set; }
+		}
+
+		public interface IDerived1
+		{
+			Guid Id { get; set; }
+			string Derived1Name { get; set; }
+		}
+
+		public interface IDerived2
+		{
+			Guid Id { get; set; }
+			string Derived1Name { get; set; }
+		}
+
+		[TestFixture]
+		public class BaseClassAndDerivedInterfaceProxiesFixture : TestCaseMappingByCode
+		{
+			private Guid _id = Guid.NewGuid();
+
+			protected override HbmMapping GetMappings()
+			{
+				var mapper = new ModelMapper();
+				mapper.Class<Base>(
+					rc =>
+					{
+						rc.Id(x => x.Id, m => m.Generator(Generators.GuidComb));
+						rc.Property(x => x.BaseName);
+					});
+
+				mapper.JoinedSubclass<Derived1Interface>(
+					rc =>
+					{
+						rc.Proxy(typeof(IDerived1));
+						rc.Property(x => x.Derived1Name);
+					});
+
+				mapper.JoinedSubclass<Derived2Interface>(
+					rc =>
+					{
+						rc.Proxy(typeof(IDerived2));
+						rc.Property(x => x.Derived2Name);
+					});
+
+				return mapper.CompileMappingForAllExplicitlyAddedEntities();
+			}
+
+			[Test]
+			public void ProxyForBaseClass()
+			{
+				using (var s = OpenSession())
+				{
+					var b = s.Load<Base>(_id);
+					Assert.That(b.Id, Is.EqualTo(_id));
+				}
+			}
+
+			[Test]
+			public void ProxyForDerived1Interface()
```

Comment:
This test is failing currently as generated proxy is not based on `Base` class

Reply:
I don't think that this test is valid. It would be extremely hard (and lead to inconsistencies) to find a _suitable_ base class for the proxy (there are several criteria: class shall be mapped, don't have an interface proxy, have all fields and methods overridable, etc). I don't think we should do it. As I said (or maybe I only thought(?)) in the issue discussion, the only concern we should fix is that the subclasses shall not affect the base class proxy. 

In this particular example if user wants a `Base` base class then they should call `Load<Base>(id)` and it will return a proxy that a base type `Base` and implements all subclass interfaces.

Reply:
Or at least not in this PR, as it is a separate issue to what @fredericDelaporte reported if I understand correctly.

Reply:
Ok. It might be hard to implement and we can skip supporting this case for this matter. But I don't see any possible inconsistencies...

Reply:
I meant that user would not be able to predict if the code is correct or not as the result would be dependant on the hierarchy and mapping and get some unexpected result. So it would be inconsistent between similar but distinct hierarchies.

Reply:
> there are several criteria: class shall be mapped, don't have an interface proxy, have all fields and methods overridable, etc

Hm. That's not the routine I have in mind.  I think we should take as a base first class specified as proxy in hierarchy and implement interfaces along the way to this class proxy.

Reply:
> I think we should take as a base first class specified as proxy in hierarchy and implement interfaces along the way to this class proxy.

I did not get it. 

Consider following hierarchy:

```csharp
//mapped
class A { public virtual int Id }
// not mapped
class B : A { public virtual int Something { get; set; } } 
// not mapped 
class C : B { public int Another {get; set;} }} 
interface ID { int Something { get; set; } int Another {get; set;} int MyProperty {get; set;} }
// mapped with proxy="ID"
class sealed D: C, ID { public int MyProperty {get; set;} } 
```

What shall the base class for proxy of `D` be?

Reply:
`C` not mapped. So it's implicitly treated as `Proxy(typeof(C))`. And that's the first proxy class in hierarchy for `D`.
And till `C` class we have only one interface proxy `ID`. So generated class should look like:
`class DProxy: C, ID"

Reply:
Was too fast :) Not mapped.. I see your point

Reply:
So to conclude: my desired proxy hierarchy is still possible and I think it might be useful in some scenarios. But it is not related to this particular fix and requires some additional logic to analyze base type mappings so maybe sometime later in the future. I've adjusted my tests accordingly.

Reply:
>but desired proxy class stays: `class DProxy: C, ID`

There is a trap: a field of `C` is not virtual and will always be null, which is undesirable.